File: merge_sort_python/merge_sort.py
"""Implementacao didatica do algoritmo Merge Sort em Python.

O algoritmo ordena a propria lista recebida, dividindo o problema em
sublistas menores e depois intercalando as partes ja ordenadas.
"""


# Junta duas metades ja ordenadas e grava o resultado na lista original.
def merge(lista, inicio, meio, fim):
    # Copia as duas metades que serao intercaladas.
    esquerda = lista[inicio:meio]
    direita = lista[meio:fim]

    i = 0
    j = 0
    k = inicio

    # Compara os menores elementos de cada metade e grava o menor na lista original.
    while i < len(esquerda) and j < len(direita):
        if esquerda[i] <= direita[j]:
            lista[k] = esquerda[i]
            i += 1
        else:
            lista[k] = direita[j]
            j += 1
        k += 1

    # Copia os elementos restantes da metade esquerda, se houver.
    while i < len(esquerda):
        lista[k] = esquerda[i]
        i += 1
        k += 1

    # Copia os elementos restantes da metade direita, se houver.
    while j < len(direita):
        lista[k] = direita[j]
        j += 1
        k += 1


# Funcao principal do Merge Sort.
# Ela divide a lista em duas partes, ordena cada metade recursivamente
# e depois chama a intercalação final.
def merge_sort(lista, inicio=0, fim=None):
    # Na primeira chamada, o algoritmo considera a lista inteira.
    if fim is None:
        fim = len(lista)

    # Uma sublista com zero ou um elemento ja esta ordenada.
    if fim - inicio <= 1:
        return

    meio = (inicio + fim) // 2

    # Ordena recursivamente as metades esquerda e direita.
    merge_sort(lista, inicio, meio)
    merge_sort(lista, meio, fim)

    # Junta as duas metades ordenadas.
    merge(lista, inicio, meio, fim)


if __name__ == "__main__":
    lista = [8, 3, 5, 1, 4, 2]

    merge_sort(lista)

    print("Lista original:", [8, 3, 5, 1, 4, 2])
    print("Lista ordenada:", lista)

File: merge_sort_C/merge_sort.c
#include <stdio.h>
#include <stdlib.h>
#include "merge_sort.h"

/* Funcao principal do Merge Sort.
   Ela divide o vetor em duas partes, ordena cada lado e depois intercala. */
void mergesort(int inicio, int fim, int v[])
{
  // Divide o vetor recursivamente ate chegar em partes de tamanho 1.
  if (inicio < fim)
  {
    int meio = (inicio + fim)/2;
    mergesort(inicio, meio, v);
    mergesort(meio+1, fim, v);
    intercala(inicio, meio, fim, v);
  }
}

/* Junta duas metades ja ordenadas em um vetor auxiliar
   e depois copia o resultado para o vetor original. */
void intercala(int inicio, int meio, int fim, int v[])
{
  // Cria um vetor auxiliar para armazenar a intercalação ordenada.
  int tamanho = fim - inicio + 1;
  int *aux = malloc(tamanho * sizeof(int));
  int inicio_v01 = inicio, inicio_v02 = meio+1, poslivre=0;

  // Compara os elementos das duas metades e copia o menor para o auxiliar.
  while (inicio_v01 <= meio && inicio_v02 <= fim)
  {
    if (v[inicio_v01] <= v[inicio_v02])  
      aux[poslivre++] = v[inicio_v01++];
    else  
      aux[poslivre++] = v[inicio_v02++];
  }

  // Copia os elementos restantes da primeira metade.
  while (inicio_v01 <= meio)  
    aux[poslivre++] = v[inicio_v01++];

  // Copia os elementos restantes da segunda metade.
  while (inicio_v02 <= fim)  
    aux[poslivre++] = v[inicio_v02++];

  // Devolve os valores ordenados para o vetor original.
  for (inicio_v01 = inicio; inicio_v01 <= fim; inicio_v01++)  
    v[inicio_v01] = aux[inicio_v01-inicio];

  free(aux);
}
